media/2021-05-12-paying-homage-to-the-compall-michael-pattern/5-scala3-opaque-types.worksheet.sc

```diff
@@ -0,0 +1,44 @@
+// Approach using Scala 3 Opaque Type Aliases
+
+object Units {
+  opaque type Kilometres = Double
+  opaque type Miles = Double
+
+  val ZeroKm: Kilometres = 0
+  val ZeroMi: Miles = 0
+
+  def kilometres(value: Double): Option[Kilometres] = if (value < 0) None else Some(value)
+  def miles(value: Double): Option[Miles] = if (value < 0) None else Some(value)
+
+  extension (km: Kilometres) {
+    def + (km2: Kilometres): Kilometres = km + km2
+  }
+
+  extension (miles: Miles) {
+    def toKm: Kilometres = miles * 1.6
+  }
+}
```

Comment:
The external interface of this implementation is a bit different from the interface in the previous [Scala 2 based] implementations: adding to `Kilometer` is implemented using via an extension method whereas in the other examples, it's a function that takes to `Kilometer` arguments

Reply:
I was being lazy. But I'll just go ahead and add extension methods to the other Scala 2 implementations, starting from the Value Class example